### PR TITLE
fix(sdk): Throw 'exit_task cannot depend on any other tasks.' error when an ExitHandler has a parameter dependent on other task

### PR DESCRIPTION
**Pull request by hbelmiro**
**Due to https://github.com/kubeflow/pipelines/pull/11005#issuecomment-2239295707, depends on:**
- https://github.com/kubeflow/pipelines/pull/10975
- https://github.com/GoogleCloudPlatform/oss-test-infra/pull/2313

Resolves https://github.com/kubeflow/pipelines/issues/10187

The issue happens when an ExitHandler has a parameter that depends on other task.

```python
from kfp import dsl
from kfp.dsl import PipelineTaskFinalStatus

@dsl.component
def test_step_a(input: str) -> str:
    return "step-a"


@dsl.component
def test_step_b(input: str, status: PipelineTaskFinalStatus):
    print("step-b")


@dsl.component
def test_step_c(input: str) -> str:
    return "step-c"


@dsl.pipeline
def test_pipeline():
    pre_task = test_step_a(input="pre")
    exit_task = test_step_b(input=pre_task.output)  <--- Here's the problem
    with dsl.ExitHandler(exit_task=exit_task):
        test_step_c(input="test")
```

If we change that line to:

```python
    exit_task = test_step_b(input="some string")
```

The pipeline will compile.

This PR aims to show the proper error message when that happens.
This may not be the right approach. I changed the code to check the ExitHandler inputs. It seems to me that https://github.com/kubeflow/pipelines/blob/f328f0b588c35cdf1e5b31638fbe3596b2f38413/sdk/python/kfp/dsl/structures.py#L417 should contain `test_step_a` though. But [I'm not sure that attribute actually means what I think it does](https://github.com/kubeflow/pipelines/blob/b57f9e858880afcbeac51a5d4e978133be6c0d50/sdk/python/kfp/compiler/compiler_utils.py#L728).

**Checklist:**
- [x] The title for your pull request (PR) should follow our title convention. [Learn more about the pull request title convention used in this repository](https://github.com/kubeflow/pipelines/blob/master/CONTRIBUTING.md#pull-request-title-convention).

### CODE REVIEW

File: sdk/python/kfp/dsl/tasks_group.py [hunk 1]
```diff
@@ -140,7 +143,16 @@ def __init__(
...
-        self.exit_task = exit_task
+    def __has_dependent_tasks(self) -> bool:
+        if self.exit_task.dependent_tasks:
+            return True
+
+        if not self.exit_task.inputs:
+            return False
+
+        return any(
+            type(task_input) is PipelineParameterChannel
```
Comment (gmfrasca) [inline sdk/python/kfp/dsl/tasks_group.py:154]: i believe it's more pythontic to use `isinstance` here instead of manually checking type

Comment (diegolovison) [reply sdk/python/kfp/dsl/tasks_group.py:154]: It depends. `type` will assert without inheritance. If it is the case, then it is fine.

Comment (hbelmiro) [reply sdk/python/kfp/dsl/tasks_group.py:154]: > `type` will assert without inheritance.

That said, I think we need `isinstance` here.

**Comment by hbelmiro** [comment]
`kubeflow-pipelines-sdk-execution-tests` should not be failing. When I run those tests on GitHub Actions, [they pass](https://github.com/kubeflow/pipelines/actions/runs/9998461907/job/27637391548?pr=11005). (I tried that by copying the workflow from https://github.com/kubeflow/pipelines/pull/10975).

**Comment by hbelmiro** [comment]
/hold

Due to https://github.com/kubeflow/pipelines/pull/11005#issuecomment-2239295707, depends on:
- https://github.com/kubeflow/pipelines/pull/10975
- https://github.com/GoogleCloudPlatform/oss-test-infra/pull/2313

**Comment by hbelmiro** [comment]
@gmfrasca @diegolovison can you guys take a second look, please?
Failing tests are due to https://github.com/kubeflow/pipelines/pull/11005#issuecomment-2239295707.

**Review by chensun — APPROVED** [review]
/lgtm
/approve

**Comment by hbelmiro** [comment]
/unhold

The depending PRs were merged.
- https://github.com/kubeflow/pipelines/pull/10975
- https://github.com/GoogleCloudPlatform/oss-test-infra/pull/2313

**Pull request merged**